### PR TITLE
FEAT-#5317: introduce `order_broken` parameter for `partition.mask`

### LINKED ISSUE #5317: FEAT: make `partition.mask` able to work with labels whose order is not the same as the order in which they are stored.
This change is necessary to continue improving the functionality for reordering rows or columns. It was required, for example, in https://github.com/modin-project/modin/pull/4880

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

In the future, we will have the opportunity to take all the rows or all the columns in an arbitrary order via `mask`.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5317 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 1]
```diff
@@ -204,8 +209,10 @@ def is_full_axis_mask(index, axis_length):
         row_labels = [row_labels] if is_scalar(row_labels) else row_labels
         col_labels = [col_labels] if is_scalar(col_labels) else col_labels
 
-        if is_full_axis_mask(row_labels, self._length_cache) and is_full_axis_mask(
-            col_labels, self._width_cache
+        if (
+            not order_broken
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition.py:213]: The only way to quickly find out if the order is broken is to ask the layer above that stores the indexes.

**Comment by anmyachev** [comment]
@YarShev could you take a look?

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 2]
```diff
@@ -171,16 +171,21 @@ def _iloc(df, row_labels, col_labels):  # noqa: RT01, PR01
         """Perform `iloc` on dataframes wrapped in partitions (helper function)."""
         return df.iloc[row_labels, col_labels]
 
-    def mask(self, row_labels, col_labels):
+    def mask(self, row_labels, col_labels, order_broken=False):
         """
         Lazily create a mask that extracts the indices provided.
 
+        By default, assumes that the order of the requested roes/cols is
+        the same as the order in which they are stored.
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/partitioning/partition.py:179]: I would rather move this to the new parameter's description

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 3]
```diff
@@ -171,16 +171,21 @@ def _iloc(df, row_labels, col_labels):  # noqa: RT01, PR01
         """Perform `iloc` on dataframes wrapped in partitions (helper function)."""
         return df.iloc[row_labels, col_labels]
 
-    def mask(self, row_labels, col_labels):
+    def mask(self, row_labels, col_labels, order_broken=False):
         """
         Lazily create a mask that extracts the indices provided.
 
+        By default, assumes that the order of the requested roes/cols is
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/partitioning/partition.py:178]:
```suggestion
        By default, assumes that the order of the requested rows/cols is
```

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 4]
```diff
@@ -171,16 +171,21 @@ def _iloc(df, row_labels, col_labels):  # noqa: RT01, PR01
...
+        the same as the order in which they are stored.
+
         Parameters
         ----------
         row_labels : list-like, slice or label
             The row labels for the rows to extract.
         col_labels : list-like, slice or label
             The column labels for the columns to extract.
+        order_broken : bool, default: False
+            If the value is True, then some optimizations are not applicable.
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/partitioning/partition.py:188]: let's elaborate on what optimizations are not applicable :)
after reading the docstring it doesn't clear at all what this parameter is doing and when you should pass it `True` or `False`

BTW, maybe some other name such as `ignore_locators_order` or something will be better?

File: modin/core/execution/ray/implementations/cudf_on_ray/partitioning/partition.py [hunk 5]
```diff
@@ -225,6 +225,8 @@ def mask(self, row_labels, col_labels):
             The row labels to extract.
         col_labels : list of hashable
             The column labels to extract.
+        order_broken : bool, default: False
+            If the value is True, then some optimizations are not applicable.
```
Comment (dchigarev) [inline modin/core/execution/ray/implementations/cudf_on_ray/partitioning/partition.py:229]: the parameter was added but unused :), however, this method also has some ground to implement it

**Comment by dchigarev** [comment]
Could you also please elaborate on why we should have this parameter? Why would we ever want to ignore the order provided by  `.mask` arguments? Why just not rework `is_full_axis_mask` so it returns False on arguments that intend to change the order?

File: modin/core/execution/ray/implementations/cudf_on_ray/partitioning/partition.py (hunk 5 above)
Comment (anmyachev) [reply modin/core/execution/ray/implementations/cudf_on_ray/partitioning/partition.py:229]: This change will be needed [here](https://github.com/modin-project/modin/pull/4880).

**Comment by anmyachev** [comment]
not actual

**Pull request closed without merging**